### ISSUE TITLE
What is the value proposition of tsup?

### ISSUE THREAD

**Issue by pastelmind** [egoist/tsup#806]
Hello! I just discovered tsup. However, I am struggling to see the benefit that tsup provides over using esbuild.

From what I gather, tsup's advantages are limited to:

- Invoke `tsc` together with esbuild, emitting type definitions (`*.d.ts`)
    - ...But I can just invoke `tsc` manually in my build scripts?
- Support CSS, SASS, etc
    - ...But these are limited to features already supported by esbuild, so why not just use esbuild directly?
- Transpile to ES5 using swc
    - ...But, if you're going to use swc anyway, why use esbuild at all?
- Zero config
    - ...But, in my experience, most projects require some build configuration anyway. Very few projects can get by with _truly_ zero config.
    - One _could_ argue that providing "sensible defaults" is the core value of tsup. If so, this should be advertised better--perhaps a comparison between a regular esbuild setup with an equivalent tsup config?

Please understand that I am not trying to attack tsup; I just want to know what core value it provides. The way I see it, tsup is very similar to Vite in that it is a mashup of different tools (esbuild, swc). However, tsup appears to lack a core feature that isn't provided by the underlying tools themselves (as opposed to Vite, which provides hot module replacement). I would be happy to be proved wrong.

Oh, and please put that value proposition statement at the top of the README.

**Comment by panoply** [comment]
Great question!

I think it mostly comes down to use case, maybe some other folks can chime in here. I'll list a few of the key features and reasonings as to why I choose tsup even though I use esbuild a fair bit.

> The way I see it, tsup is very similar to Vite in that it is a mashup of different tools (esbuild, swc).

I'd consider tsup a complete different tool and the appropriation to differ even though I'm not a Vite fan nor a user (the subtle Vue branding imposed by the tool is insufferable) but opinions aside, it's my understanding that Vite tends to be geared toward web projects, specifically SPA's. You'll likely see tsup used for JavaScript modules more than web oriented projects. Personally I reach for tsup when bundling things like parsers, api clients and more JavaScript focused tooling but in saying that I do indeed leverage it in SPAs likely because my preferred stack choices in that area make tsup the perfect fit but grain of salt this claim as I am basing it from my own usage and observations.

> Invoke tsc together with esbuild, emitting type definitions (*.d.ts) ...But I can just invoke tsc manually in my build scripts?

If only it was that easy... 

> - Transpile to ES5 using swc
>   - ...But, if you're going to use swc anyway, why use esbuild at all?

Definitely a valid argument to be had here but I'd be willing to bet that very few developers even know about SWC. I suppose it comes down to a matter of convenience, personally I'd more readily pull in tsup over SWC if I needed ES5. 

> One could argue that providing "sensible defaults" is the core value of tsup. If so, this should be advertised better--perhaps a comparison between a regular esbuild setup with an equivalent tsup config?

In addition to the CLI, consumers of this module have a multiple different methods to choose from when defining build configuration. Options can be defined in `tsup.config.ts`, `tsup.config.js`, `tsup.config.json` or within package.json files via `tsup` key. The configuration structure is refined in tsup and this alleviates preset delineation by exposing build options relevant for most modern projects. In esbuild the granular set of options can sometimes make adoption and implementation a little difficult for developers who and when choosing tsup one does not need to think or dig deep to produce distributed code of quality.

Another key selling point here is the ability to effortlessly have tsup produce multiple bundles from a single configuration. While you can do this in esbuild, it is just far easier to do with tsup while still being able to leverage the additional capabilities offered like declaration generation, splitting, tree-shaking etc.


- **Code Splitting / Tree Shaking and Rollup**
  - Though considered "experimental" the code splitting capabilities of tsup are essential to projects which require it and tree-shaking is another key capability of this nexus, both of which are supported by tsup and not esbuild. The beauty of tsup is that couples the best known solutions together in an elegant manner providing you the speed and power of esbuild but also the fastidious processing of Rollup. 

- **Easy Adoption**
  - Before Evan bestowed upon us the gift that is esbuild working with TypeScript was rather painful. A lot of developers who were using Rollup (for example) would need to rely on plugins for transpiling code and the entire process was slow, error prone. I actually know of case where a company relying heavily on the generated output produced by Rollup had to initially forgo esbuild adoption due to the undertaking involved in refining esbuild to replicating their derivation. They actually end up using tsup because it allowed them to produce near identical results to that they were getting with Rollup without needing to extend upon the existing capabilities due to tsup running code through rollup in its post-process cycle.
     
- **Egoist**
  - I like to support talented developers who create brilliant open source tools. Egoist is one of the developers. 
  
While you make more than valid arguments, all of which I actually agree with you on, I think you have a deeper understanding of the ecosystem and build tooling as a whole. You are coming from a standpoint of experience and maybe forgetting that most devs have trouble configuring a `tsconfig.json` file let alone retain fluency in syntax availability of EcmaScript releases.

**Comment by pastelmind** [comment]
@panoply Thanks for your detailed and thoughtful response. It made me warm up a bit towards tsup.

I read the docs more thoroughly and discovered some tsup features that make sense for building libraries:

- All dependencies are marked as external by default (i.e. not bundled, but imported at runtime)
- Can inject shims for `__dirname` in ESM output, and `import.meta.url` in CJS output

These small perks may add up to a nontrivial improvement in overall developer experience. Though I still believe tsup fits a less-than-general case--for people who maintain many libraries and want a simplified solution--I'm starting to see what it's trying to do.

**Comment by datner** [comment]
@pastelmind sorry to necro this, but it's a pet peeve of mine..
Not really sure what is the point of this issue. This reductionist approach can be applied to discount the value of any abstraction..
If `tsup` brings you the same value at the same cost as vanilla `esbuild`, then just use `esbuild`. 

But the biggest misunderstanding is in the statement "tsup fits a less-than-general case--for people who maintain many libraries". Au contraire, your initial comparison with `vite` was a much more apt observation. `tsup` is great for the general case. It's great for non-maintainers, people who do not have massive bundler experience in the most egregious of conditions, it's great for most projects. It being consistent and opinionated gives it the same exact advantage that allowed `vite` to dominate `snowpack`  so hard it went out of the game. It's value _is_ that it Just Works™️ . `vite` having `hmr` is a great boon, but it's not why it's so popular. It's the fact that with a very simple `vite.config.ts` and a couple plugins you start running and you don't feel the burn until you're at **massive** scale. Unmatched experience to previous tools.

`tsup` remedies the same pains but in node (and modules, therefore libs) instead. Though js/ts backend people don't tend to like comfortable or modern tooling, they want to roll with their frozen-version legacy framework builtins, and frontend people have `vite` so I unfortunately do not foresee a similar future to `vite`...

TL;DR
if you're making a node server or lib use `tsup`. It's built for the general case, not the special case. Save time, have fun.

**Comment by fwextensions** [comment]
I had the same question, so adding to this issue instead of starting a new one.  The value prop from the homepage is:

> Bundle your TypeScript library with no config

But this didn't actually work for me.  I'd previously cobbled together a build process for a simple library using Rollup, but I wanted to try tsup with a new, extremely simple library that contained just two files and an `index.ts` that re-exported the other files. 

Running `tsup` alone just showed an error that there were no input files.  So I tried `tsup src`, thinking it would bundle everything in the `src/` directory.  But it just produced .cjs versions of all the files, with no .d.ts files (which I thought was the whole point?).  

Then I tried adding some config to the package.json:

```json
  "tsup": {
    "entry": ["src/index.ts"],
    "format": ["esm"],
    "dts": true,
    "sourcemap": true,
    "clean": true,
    "minify": false
  },
```

This at least generated .d.ts files, but they weren't bundled together, and it generated all these extra chunk files, which again seemed like the opposite of bundling. 

After some more fumbling, I tried calling `tsup src/index.ts` (even though that's specified as the entry in the config).  Finally, just one bundled `index.js` file and one bundled `index.d.ts` file were generated.  

Now, maybe I'm just too much of a newbie to use tsup, but I have generated TypeScript npm packages with Rollup before, so I'm not completely new to this.  I think the docs could help clarify how to bundle your library files into a single .d.ts file, which is likely why you're turning to this tool in the first place.
